### PR TITLE
Prevent possible infinite recursion

### LINKED ISSUE #2127: `final_exception': stack level too deep (SystemStackError)
This error is happening in all rspec-rails 3.x versions I tested here. since 3.0 until 3.4

`/Users/luciano/.rvm/gems/ruby-2.2.3/gems/rspec-core-3.4.1/lib/rspec/core/formatters/exception_presenter.rb:97:in `final_exception': stack level too deep (SystemStackError)`
https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters/exception_presenter.rb#L97

note my ruby version. 2.2.3.
I teste in the ruby 2.1.3 and happen the same.

### PR DESCRIPTION
**Pull request by JonRowe**
Fixes #2127

### CODE REVIEW

**Comment by seanwalbran** [comment]
@JonRowe Just a note that this doesn't catch cause loops of more than one exception (e.g. when `error.cause != error`, but `error.cause.cause == error`), which I happened to run into with rails+haml: https://gist.github.com/seanwalbran/b63fbe7117645c5ddef1#file-gistfile1-txt-L26

**Comment by myronmarston** [comment]
Thanks for reporting that, @seanwalbran!

It sounds like the fix needs to be more sophisticated -- either it needs to keep track of all exceptions seen so far (e.g. via an extra arg passed as part of the recursion) so it can notice when it's seeing one it has already seeing, or maybe we should just put a hard limit on the max depth (e.g. 10 or 20 or something) and give up when we've followed `Exception#cause` that many times.

**Comment by JonRowe** [comment]
I've reworked this to take @seanwalbran's issue into account

File: lib/rspec/core/formatters/exception_presenter.rb [hunk 1]
```diff
@@ -92,9 +92,10 @@ def fully_formatted_lines(failure_number, colorizer)
 
       private
 
-        def final_exception(exception)
-          if exception.cause
-            final_exception(exception.cause)
+        def final_exception(exception, previous=[])
+          if exception.cause && !previous.include?(exception.cause)
+            previous << exception.cause
+            final_exception(exception.cause, previous)
```
Comment (myronmarston) [inline lib/rspec/core/formatters/exception_presenter.rb]: It'll probably be slightly faster if you call `cause` only once:

``` ruby
def final_exception(exception, previous=[])
  cause = exception.cause
  return exception unless cause
  return exception if previous.include?(cause)

  previous << cause
  final_exception(cause, previous)
end
```

Not a merge blocker, though.

**Comment by myronmarston** [comment]
LGTM otherwise.

**Comment by joeletizia** [comment]
Any chance of a 3.5 release for this this?

**Comment by JonRowe** [comment]
We could put out a patch release of `3.4` (@samphippen might need your help)

**Comment by taylorzr** [comment]
I ran into this issue as well, any chance this will be released as a patch to 3.4?

**Comment by JonRowe** [comment]
Hmm apparently I forgot to backport this so it was never released in recent patch versions, something I've just rectified with 3.4.3.

**Comment by dgilperez** [comment]
I just got hit by this at 3.4.4 - expected?

**Comment by JonRowe** [comment]
This should have been rectified in `3.4.4` can you provide more details?

**Comment by dgilperez** [comment]
I'm sorry, I somehow worked around it and can't reproduce it now.

**Pull request merged**